search/localsearch_nqueens.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * @file
+ * @brief [Heuristic local Search](https://en.wikipedia.org/wiki/Local_search_(optimization)) algorithm.
+ * @details
+ * The LS algorithm is the basis of many methods used for optimization problems.
+ * Its system is based on the search for neighborhoods that obtain a better solution.
+ * This exercise solves the n queens problem [https://en.wikipedia.org/wiki/Eight_queens_puzzle].
+ * The rule of neighborhood applied in this file is first.improvement.
+ * @author [dhernandezl](https://github.com/dhernandezl)
+ */
+
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include <ctime>
+
+#define NREINAS 8 //The number of queens on the table is the same as the dimension of the chessboard.
+				  // NREINAS 4; NREINAS 8; NREINAS 16
+/**
+ * function: initial_solution
+ * Initialize a new solution, generating
+ * random numbers with the maximum height of the board.
+ * @returns a vector that will be the starting solution.
+ */
+std::vector<int> initial_solution() {
+	std::vector<int> solution;
+	srand(time(nullptr)); //reset srand
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		solution.push_back(rand() % NREINAS + 1);
+	}
+	return solution;
+}
+
+/**
+ * function: cost
+ * Calculate the cost of the solution vector,
+ * the optimal cost is h = 1. Read the 8 queens problem.

```

Comment:
Isn't the optimal cost = 0? We're looking for solution such that no two queens are "attacking" each other.

Reply:
Yes, exactly, but this is a heuristic algorithm that is the basis of many algorithms. The algorithm literature places it as the minimum to reach, because the local search can fall into local optima.